src/pyiterator.jl

```diff
@@ -29,16 +30,81 @@ end
 
     Base.done(po::PyObject, s) = ispynull(s[1])
 else
-    function Base.iterate(po::PyObject, s=_start(po))
+    """
+        PyIterator{T}(pyobject)
+
+    Wrap `pyobject::PyObject` into an iterator, that produces items of type `T`. To be more precise `convert(T, item)` is applied in each iteration. This can be useful to avoid automatic conversion of items into corresponding julia types.
+    ```jldoctest
+    julia> using PyCall
+
+    julia> l = PyObject([PyObject(1), PyObject(2)])
+    PyObject [1, 2]
+
+    julia> piter = PyCall.PyIterator{PyAny}(l)
+    PyCall.PyIterator{PyAny,Base.HasLength()}(PyObject [1, 2])
+
+    julia> collect(piter)
+    2-element Array{Any,1}:
+     1
+     2
+
+    julia> piter = PyCall.PyIterator(l)
+    PyCall.PyIterator{PyObject,Base.HasLength()}(PyObject [1, 2])
+
+    julia> collect(piter)
+    2-element Array{PyObject,1}:
+     PyObject 1
+     PyObject 2
+    ```
+    """
+    struct PyIterator{T,S}
+        o::PyObject
+    end
+
+    function _compute_IteratorSize(o::PyObject)
+        S = try
+            length(o)
+            Base.HasLength
+        catch err
+            if !(err isa PyError && pyisinstance(err.val, @pyglobalobjptr :PyExc_TypeError))
+                rethrow()
+            end
+            Base.SizeUnknown
+        end
+    end
+    function PyIterator(o::PyObject) 
+        PyIterator{PyObject}(o)
+    end
+    function (::Type{PyIterator{T}})(o::PyObject) where {T}
+        S = _compute_IteratorSize(o)
+        PyIterator{T,S}(o)
+    end
+    
+    Base.eltype(::Type{<:PyIterator{T}}) where T = T
+    Base.eltype(::Type{<:PyIterator{PyAny}}) = Any
+    Base.length(piter::PyIterator) = length(piter.o)
+
+    Base.IteratorSize(::Type{<: PyIterator{T,S}}) where {T,S} = S()
+
+    _start(piter::PyIterator) = _start(piter.o)
+    
+    function Base.iterate(piter::PyIterator{T}, s=_start(piter)) where {T}
         ispynull(s[1]) && return nothing
         sigatomic_begin()
         try
             nxt = PyObject(@pycheck ccall((@pysym :PyIter_Next), PyPtr, (PyPtr,), s[2]))
-            return (convert(PyAny, s[1]), (nxt, s[2]))
+            return (convert(T,s[1]), (nxt, s[2]))
         finally
             sigatomic_end()
         end
     end
+    function Base.iterate(po::PyObject, s=_start(po))
+        # avoid the constructor that calls length
+        # since that might be an expensive operation
+        # even if length is cheap, this adds 10% performance
+        piter = PyIterator{PyAny, Base.IteratorSize}(po)
```

Comment:
Good catch :)

Reply:
Do you mean to pass `Base.SizeUnkown` as the second parameter?

Reply:
Not really, the value of the second parameter is completely irrelevant.

Reply:
Isn't the second parameter used as the return value for `Base.IteratorSize(::Type{<: PyIterator{T,S}})`?

Reply:
In general sure. But the specific object we are talking about here only lives during a single `iterate` call and hence there is no chance to call `Base.IteratorSize` on it.

Reply:
@jw3126 I thought `Base.IteratorSize` was OK since I realized that this type parameter won't be used anywhere.  But I think there was a few seconds of what??? moment.  So maybe just use `Base.SizeUnkown` and comment that this type parameter will not be used?  Using `Base.SizeUnkown` could also do a super micro-optimization (nano-optimization?) to reduce number of methods (nah, I guess not :smiley: ).

Reply:
I did the change, I also think it is clearer now. Also consider, that `Base.SizeUnkown` has less characters, so we can save lots of parse time in addition :smile:.